pkg/server/grpc_server.go

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) 2022 NTT Communications Corporation
+//
+// This software is released under the MIT License.
+// see https://github.com/nttcom/pola/blob/main/LICENSE
+
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	pb "github.com/nttcom/pola/api/grpc"
+	"github.com/nttcom/pola/internal/pkg/cspf"
+	"github.com/nttcom/pola/internal/pkg/table"
+	"github.com/nttcom/pola/pkg/packet/pcep"
+	"go.uber.org/zap"
+	grpc "google.golang.org/grpc"
+)
+
+type APIServer struct {
+	pce        *Server
+	grpcServer *grpc.Server
+	pb.UnimplementedPceServiceServer
+}
+
+func NewAPIServer(pce *Server, grpcServer *grpc.Server) *APIServer {
+	s := &APIServer{
+		pce:        pce,
+		grpcServer: grpcServer,
+	}
+	pb.RegisterPceServiceServer(grpcServer, s)
+
+	return s
+}
+
+func (s *APIServer) Serve(address string, port string) error {
+	var listenInfo strings.Builder
+	listenInfo.WriteString(address)
+	listenInfo.WriteString(":")
+	listenInfo.WriteString(port)
+	s.pce.logger.Info("gRPC Listen", zap.String("listenInfo", listenInfo.String()), zap.String("server", "grpc"))
+	grpcListener, err := net.Listen("tcp", listenInfo.String())
+	if err != nil {
+		return err
+	}
+
+	if err := s.grpcServer.Serve(grpcListener); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *APIServer) CreateSrPolicy(ctx context.Context, input *pb.CreateSrPolicyInput) (*pb.SrPolicyStatus, error) {
+	if s.pce.ted == nil {
+		return &pb.SrPolicyStatus{IsSuccess: false}, errors.New("ted is disable")
+	}
+
+	// Validate input
+	if input.GetSrPolicy().GetPcepSessionAddr() == nil || input.GetAsn() == 0 || input.GetSrPolicy().GetColor() == 0 || input.GetSrPolicy().GetSrcRouterId() == "" || input.GetSrPolicy().GetDstRouterId() == "" {
+		return &pb.SrPolicyStatus{IsSuccess: false}, errors.New("input is invalid")
+	}
+
+	s.pce.logger.Info("Get request CreateSrPolicy API", zap.Any("SR Policy", input.GetSrPolicy()), zap.String("server", "grpc"))
```

Comment:
Modify "pcepSessionAddr", "LoAddr", etc. to read as strings.

```
2022-10-11T12:25:04.448Z	info	Get request CreateSrPolicy API	{"SR Policy": "pcepSessionAddr:\"\\nd\\x00\\x01\"  srcRouterId:\"0000.0aff.0001\"  dstRouterId:\"0000.0aff.0004\"  color:100  policyName:\"policy-name\"  type:DYNAMIC  metric:TE", "server": "grpc"}
2022-10-11T12:25:04.449Z	info	Send PCInitiate	{"session": "10.100.0.1", "srpId": 1, "policyName": "policy-name", "labels": [{"Sid":16002,"LoAddr":"Cv8AAg=="},{"Sid":16004,"LoAddr":"Cv8ABA=="}], "color": 100, "preference": 100, "srcIPv4": "Cv8AAQ==", "dstIPv4": "Cv8ABA=="}
```

Reply:
```
2022-10-17T09:09:57.608Z	info	Get request CreateSrPolicy API	{"input": {"asn":"65000","srPolicy":{"color":200,"dstRouterId":"0000.0aff.0004","metric":"IGP","pcepSessionAddr":"10.100.0.1","segmentList":[{"sid":16002},{"sid":16003},{"sid":16004}],"srcRouterId":"0000.0aff.0001","type":"EXPLICIT"}}, "server": "grpc"}
2022-10-17T09:09:57.608Z	info	Send PCInitiate	{"session": "10.100.0.1", "srpId": 1, "policyName": "juno-tmp2", "labels": [{"LoAddr":"10.255.0.2","Sid":16002},{"LoAddr":"10.255.0.3","Sid":16003},{"LoAddr":"10.255.0.4","Sid":16004}], "color": 200, "preference": 100, "srcIPv4": "10.255.0.1", "dstIPv4": "10.255.0.4"}
```